### ISSUE TITLE
Introduce registered stakepools early in the Shelley era in Cardano ThreadNet tests

### ISSUE THREAD

**Issue by nfrisby** [input-output-hk/ouroboros-network#2388]
At the moment, the tests fail with a (de)centralization parameter of `d=0`, because at the moment, only the overlay slots of the Shelley era have leaders.

**Comment by nfrisby** [comment]
Some related dialogue on `ouroboros-prototyping` on Slack https://input-output-rnd.slack.com/archives/GCW2S19CM/p1594392483235100

**Comment by edsko** [comment]
Lowering to medium priority as we could potentially release without. Assigned to current milestone though.

**Comment by nfrisby** [comment]
Blocked by input-output-hk/cardano-ledger-specs#1721 -- the ThreadNet tests that involve Shelley use `ConcreteCrypto`.

https://github.com/input-output-hk/ouroboros-network/blob/bcd7a88beb6d5d30558aed00081855d256eea0c1/ouroboros-consensus-shelley-test/src/Test/Consensus/Shelley/MockCrypto.hs#L15-L21

ConcreteCrypto assumes ADDR and ADDRHASH are equivalent.

https://github.com/input-output-hk/cardano-ledger-specs/blob/0bea2d30be6290c92dece2e1d58d03165832219b/shelley/chain-and-ledger/executable-spec/test/Test/Shelley/Spec/Ledger/ConcreteCryptoTypes.hs#L68-L70

Moreover, so does some of the relevant code in `cardano-ledger-specs`

```
nfrisby@frisbycomp:~/cardano-ledger-specs$ git grep -e '\<HASH\>.*\<ADDRHASH\>' -e '\<ADDRHASH\>.*\<HASH\>' -- '*hs'
shelley/chain-and-ledger/executable-spec/test/Test/Shelley/Spec/Ledger/ConcreteCryptoTypes.hs:    HASH c ~ ADDRHASH c
shelley/chain-and-ledger/executable-spec/test/Test/Shelley/Spec/Ledger/Serialisation/Generators.hs:import Shelley.Spec.Ledger.Crypto (ADDRHASH, Crypto, HASH)
shelley/chain-and-ledger/executable-spec/test/Test/Shelley/Spec/Ledger/Serialisation/Generators.hs:    HASH c ~ ADDRHASH c
nfrisby@frisbycomp:~/cardano-ledger-specs$ git rev-parse HEAD
83d8eaa9773e207491ff62dd59a8714cdb24a6d7
```

This is problematic because it prevents us from using different hashes for blocks and addresses in the Shelley tests. Byron uses Blake2b_224 for addresses and Blake2b_256 for hashes. Our Shelley tests are currently using `Blake2b_256` for both. Unfortunately, the Byron-to-Shelley translation function assumes that both Byron and Shelley are using the same hash for addresses. Because they don't in the tests, I'm seeing `error "bootstrapKeyHash: incorrect hash length"`. (The HFC (currently?) assumes that the block hash (size?) remains the same, so we can't just switch all of Shelley to 224 in the tests. My naive attempt at that gave `error "differing values across hard fork"`.)

Once the Cardano ThreadNet tests reach the Shelley era, all of the funds are tied up in Byron bootstrap addresses. Those can't delegate stake to pools, so we will have to relocate the funds to non-bootstrap (Shelley) addresses as a first step. This requires those bootstrap addresses to be usable -- that's how all of the above relates to this Issue.

**Comment by nfrisby** [comment]
Update: still blocked. See Issue #2488 for more details (blocked on a `cardano-ledger-specs` Issue that is blocked on another `ouroboros-network` Issue 😵).

**Comment by nfrisby** [comment]
I have a branch that includes a workaround to the blocking Issue 2488, [nfrisby/issue-2388-stakepools-Cardano-ThreadNet](https://github.com/input-output-hk/ouroboros-network/tree/nfrisby/issue-2388-stakepools-Cardano-ThreadNet) = 783fa3bed. It uses `Ed25519DSIGN` in the Cardano ThreadNet tests to circumvent Issue 2488. I'd rather use as much mock crypto as possible, but this can be temporary. That branch depends on PR #2490, which brings in necessary upstream changes in `cardano-ledger-specs`.

I believe my branch is ready for review, once PR #2490 is merged. It will resolve this Issue, and open up a follow-up Issue that is blocked by Issue 2488 to return to `MockDSIGN`.
